ot/gromov/_unbalanced.py

```diff
@@ -0,0 +1,1210 @@
+            ucoot : float
+                Total cost.
+    """
+
+    #############################
+    # Calculate D(pi, a \otimes b) for D = KL and squared L2.
+
+    # Calculate KL(pi, a \otimes b) using
+    # the marginal distributions pi1, pi2 of pi
+    def approx_shortcut_kl(pi, pi1, pi2, a, b):
+        """
+        Implement:
+        < pi, log pi / (a \otimes b) >
+        = <pi, log pi> - <pi1, log a> - <pi2, log b>.
+        """
+
+        res = nx.sum(pi * nx.log(pi + 1.0 * (pi == 0))) \
+            - nx.sum(pi1 * nx.log(a)) - nx.sum(pi2 * nx.log(b))
+        return res
+
+    def div(pi, pi1, pi2, a, b):
+        """
+        Calculate D(pi, a \otimes b).
+        """
+
+        if divergence == "kl":
+            res = approx_shortcut_kl(pi, pi1, pi2, a, b) \
+                - nx.sum(pi1) + nx.sum(a) * nx.sum(b)
+        elif divergence == "l2":
+            res = (nx.sum(pi**2) + nx.sum(a**2) * nx.sum(b**2)
+                   - 2 * nx.dot(a, pi @ b)) / 2
+        return res
+
+    #############################
+    # Support functions for KL and squared L2 between product measures:
+    # Calculate D(mu \otimes nu, alpha \otimes beta).
+    def approx_kl(p, q):
+        return nx.sum(p * nx.log(p + 1.0 * (p == 0))) - nx.sum(p * nx.log(q))
+
+    def kl(p, q):
+        return approx_kl(p, q) - nx.sum(p) + nx.sum(q)
+
+    def product_kl(mu, nu, alpha, beta):
+        """
+        Calculate the KL divergence between two product measures:
+        KL(mu \otimes nu, alpha \otimes beta) =
+        m_mu * KL(nu, beta) + m_nu * KL(mu, alpha) +
+        (m_mu - m_alpha) * (m_nu - m_beta)
+
+        Parameters
+        ----------
+        mu: vector or matrix
+        nu: vector or matrix
+        alpha: vector or matrix with the same size as mu
+        beta: vector or matrix with the same size as nu
+
+        Returns
+        ----------
+        KL divergence between two product measures
+        """
+
+        m_mu, m_nu = nx.sum(mu), nx.sum(nu)
+        m_alpha, m_beta = nx.sum(alpha), nx.sum(beta)
+        const = (m_mu - m_alpha) * (m_nu - m_beta)
+        res = m_nu * kl(mu, alpha) + m_mu * kl(nu, beta) + const
+
+        return res
+
+    def product_l2(mu, nu, alpha, beta):
+        """
+        norm = ||mu \otimes nu - alpha \otimes beta ||^2
+        = ||a||^2 ||b||^2 + ||mu||^2 ||nu||^2 - 2 < alpha, mu > < beta, nu >.
+        L2(mu \otimes nu, alpha \otimes beta) = norm / 2.
+        """
+        norm = nx.sum(alpha**2) * nx.sum(beta**2) \
+            - 2 * nx.sum(alpha * mu) * nx.sum(beta * nu) \
+            + nx.sum(mu**2) * nx.sum(nu**2)
+
+        return norm / 2
+
+    def product_div(mu, nu, alpha, beta):
+        if divergence == "kl":
+            return product_kl(mu, nu, alpha, beta)
+        elif divergence == "l2":
+            return product_l2(mu, nu, alpha, beta)
+
+    #############################
+    # Support functions for BCD schemes
+    def local_cost(data, pi, tuple_p, hyperparams):
```

Comment:
would that be useful to a user that wants to compute the loss out of the solver ? 
If it is the case, the function could be defined out of this wrapper. If not, it seems that some inputs to the function won't be change while running the solver so these parameters could be removed from the inputs for readibility.

Reply:
Noted. Now move to _utils.py